firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java

```diff
@@ -203,92 +213,90 @@ public synchronized void unregisterActivityLifecycleCallbacks() {
+   * classes. Fallback to class-load time of a Firebase class below API 24.
    *
    * @return {@link Timer} at the beginning of app start by Fireperf definition.
    */
-  private static Timer getStartTimer() {
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
-      return Timer.ofElapsedRealtime(
-          Process.getStartElapsedRealtime(), Process.getStartUptimeMillis());
+  private Timer getStartTimer() {
+    // Preferred: Android system API provides BIND_APPLICATION time
+    if (processStartTime != null) {
+      return processStartTime;
+    }
+    // Fallback: static initializer time (during class-load) of a Firebase class
+    return getClassLoadTime();
+  }
+
+  /**
+   * Timestamp during class load. This timestamp is captured in a static initializer during class
+   * loading of a particular Firebase class, the order of which CANNOT be guaranteed to be prior to
+   * all other classes of an app, nor prior to resource-loading of an app. Thus this timestamp is
+   * NOT PREFERRED to be used as starting-point of app-start.
+   *
+   * @return {@link Timer} captured by static-initializer during class-loading of a Firebase class.
+   */
+  private Timer getClassLoadTime() {
+    // Prefered: static-initializer time of the 1st Firebase class during init
+    if (firebaseClassLoadTime != null) {
+      return firebaseClassLoadTime;
     }
-    return firebaseStartupTime;
+    // Fallback: static-initializer time of the current class
+    return PERF_CLASS_LOAD_TIME;
   }
 
-  private void recordFirstDrawDone() {
-    if (firstDrawDone != null) {
+  private void recordPreDraw() {
+    if (preDrawPostTime != null) {
       return;
     }
-    Timer start = getStartTimer();
-    this.firstDrawDone = clock.getTime();
+    this.preDrawPostTime = clock.getTime();
     this.experimentTtid
-        .setClientStartTimeUs(start.getMicros())
-        .setDurationUs(start.getDurationMicros(this.firstDrawDone));
-
-    TraceMetric.Builder subtrace =
-        TraceMetric.newBuilder()
-            .setName("_experiment_classLoadTime")
-            .setClientStartTimeUs(firebaseStartupTime.getMicros())
-            .setDurationUs(firebaseStartupTime.getDurationMicros(this.firstDrawDone));
-    this.experimentTtid.addSubtraces(subtrace.build());
-
-    subtrace = TraceMetric.newBuilder();
-    subtrace
-        .setName("_experiment_uptimeMillis")
-        .setClientStartTimeUs(start.getMicros())
-        .setDurationUs(start.getDurationUptimeMicros(this.firstDrawDone));
-    this.experimentTtid.addSubtraces(subtrace.build());
-
-    this.experimentTtid.addPerfSessions(this.startSession.build());
-
-    if (isExperimentTraceDone()) {
-      executorService.execute(() -> this.logExperimentTtid(this.experimentTtid));
+        .setClientStartTimeUs(getStartTimer().getMicros())
+        .setDurationUs(getStartTimer().getDurationMicros(this.preDrawPostTime));
+    logExperimentTrace(this.experimentTtid);
```

Comment:
We are logging logExperimentTrace from two places - `recordPreDraw` and `recordPreDrawFrontOfQueue` (line 299). Are we logging events twice?

Reply:
`logExperimentTrace` should start with a check that won't send until all 3 traces are captured. The reason being that there's no guarantee which one will finish first for draw vs predraw, I can explain better in person.